pkg/webutil/middleware.go

```diff
@@ -0,0 +1,61 @@
+package webutil
+
+import (
+	"fmt"
+	"net/http"
+)
+
+// A Middleware is used to execute intermediate Handlers
+// in response to a request.
+type Middleware func(http.Handler) http.Handler
+
+//  Chain recursively chains middleware together.
+func Chain(h http.Handler, m ...Middleware) http.Handler {
+	if len(m) == 0 {
+		return h
+	}
+	return m[0](Chain(h, m[1:cap(m)]...))
+}
+
+const (
+	HeaderAPIKey = "X-modprox-api-key"
+)
+
+// KeyGuard creates a Middleware which protects access to a handler
+// by first checking for the X-modprox-api-key header being set. If
+// at least one of the values for the header matches one of the keys
+// configured for the KeyGuard, the handler is executed for the request.
+// Otherwise, a StatusForbidden response is returned.
+func KeyGuard(keys []string) Middleware {
+	allowedKeys := make(map[string]bool)
+	for _, key := range keys {
+		allowedKeys[key] = true
+	}
+
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// first check that the header is set
+			keys, exists := r.Header[http.CanonicalHeaderKey(HeaderAPIKey)]
+			if !exists {
+				msg := fmt.Sprintf("header %s is not set in request", HeaderAPIKey)
+				http.Error(w, msg, http.StatusForbidden)
+				return
+			}
+
+			// check all the given values for the header
+			for _, key := range keys {
```

Comment:
Is there a good reason to allow clients (not the server) to specify multiple keys in any given request?

I could imagine a malicious user setting a large number of `x-modprox-api-key` headers and forcing us to iterate through all of them before denying the request.  Depending our HTTP server settings, this may contribute to a denial of service (depending on how many headers the attacker can fit into one request before we return `StatusRequestEntityTooLarge`).  I tested locally and 1 million incorrect API key headers set in a request can add about 11ms overhead to each request.

Reply:
That's a good point - in the context of rotating keys (what this was intended for) it should be fine to just let the server accept more than one key - then have the clients do the upgrade  - then have the server get rid of the old key. I'll get rid of this multiple keys on the client side stuff.